### DISCUSSION TITLE
Binding of schema to/from Python class

### DISCUSSION THREAD

**Discussion by marscher** [BAMWelDX/weldx#397] [Ideas]
Having to access a deserialized asdf file via dictionary keys works fine, but is not especially user/programmer friendly, as there is no auto-completion and documentation (like of properties/accessor methods)

For example:
``` python
process = data["process"]
process["welding_process"]
...
```

Additionally if the schema changes, scripts accessing the dictionary directly are likely to fail. 
So we should think of wrapping the schemas in Python classes/objects to avoid breaking compatibility on schema changes. There are automated ways of doing so, but we should discuss, what features are desirable before sticking to a library. We can even decide of doing it manually, but this would introduce another maintenance job keeping the schemas and classes in sync.

**Comment by CagtayFabry** [comment]
any tools/libraries you know that might be helpful?
I think in most cases creating a simple `dataclass` from an ASDF schema should be sufficient (and rather straight forward with basic features), this could be done at runtime or generated upon changes/releases

**Comment by marscher** [comment]
https://stackoverflow.com/questions/6578986/how-to-convert-json-data-into-a-python-object I assume the same thing is possible for YAML. However doing it on the fly has the downside, that it has no documentation.

**Comment by CagtayFabry** [comment]
I guess YAML/ASDF -> Python dataclass for our use case is simple enough, with some ASDF tools we should be able to quickly generate quite clean code files

An older jinja template can be found [here ](https://github.com/BAMWelDX/weldx/blob/0.2.2/weldx/asdf/templates/asdf_dataclass.py.jinja)
with some of the new ASDF tools this might be worth revisiting, certainly a useful tool for sometime in the future (I image a simple web interface/editor -> ASDF schema -> python class to be very user friendly)

If we just want to have .attr acces something simpler like your link might suffice

Maybe this is relevant as well? https://github.com/konradhalas/dacite

**Comment by CagtayFabry** [comment]
also just leaving this here for reference https://stackoverflow.com/questions/53376099/python-dataclass-from-a-nested-dict

**Comment by CagtayFabry** [comment]
if we just want to have an attribute like access styles maybe [attrdict](https://pypi.org/project/attrdict/) is handy

**Comment by CagtayFabry** [comment]
I think `attrdict` is raining the following `DeprecationWarning` in tests:
```
weldx/asdf/file.py::weldx.asdf.file.WeldxFile.as_attr
  /opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/attrdict/mixins.py:5: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    from collections import Mapping, MutableMapping, Sequence
```

should be an easy fix but unfortunately `attrdict` is set to in `inactive` https://pypi.org/project/attrdict/ and the [GitHub repo](https://github.com/bcj/AttrDict) is even archived

We should probably look for alternatives before relying too much on it? @marscher 
So far it is not used in "production" code, only as utility afaik?

**Reply by marscher** [reply]
yes, it is only a quick shortcut. It is not used internally (and most likely not outside of it). I just replaced the code with simple custom implementation. It assumes (like attrdict did), that all dict keys are valid attribute names.

**Reply by CagtayFabry** [reply]
cool, that should be more than enough 👍
